### ISSUE TITLE
Improvement of retained file selection vs. two commit diff / merge diff

### ISSUE THREAD

**Issue by mstv** [gitextensions/gitextensions#7907]
## Current behaviour

The combination of #7798 and of the #4154 implementation selects files in these additional diff groups. This can be unexpected and misleading, especially if the diff group headings are not visible due to many files.

The selection jumps to the second diff group of a merge diff when the previous commit was the left parent.

I guess the match is based only on the file name and ignores the path.

## Expected behaviour

Only matches in the first diff group are reselected.
The file path must match.

## Steps to reproduce

- select a merge commit, e.g. 37f4f9326da1ce72dafb468ab42a86e4d1eda947
- Ctrl+P, Ctrl+N

## Environment

- Git Extensions 4.0.0.8859
- Build 3bb59459d39c3fe086da40cf59994eb775562014 (Dirty)
- Git 2.24.1.windows.2
- Microsoft Windows NT 6.2.9200.0
- .NET Framework 4.8.4121.0
- DPI 96dpi (no scaling)

**Comment by gerhardol** [comment]
Selection to retain is for the complete file list, there is no method to check for a specific group right now
If there is, how should the group match? Only retain if the file exists in first group, also if another group was selected before or try to set 'classes' for the groups? If so what? (normal diff, multi selected diff, other parents, unique BASE-selected, unique BASE-first, common BASE-selected/first)
It can be quite complicated, I rather keep as is, this simplifies when switching commits.

**Comment by mstv** [comment]
Yep: "Expected behaviour: Only matches in the first diff group are reselected."
Follow the master branch from old to new (chain of merged pull requests). It selects the uninteresting group very often (always unless it is a merge which could have been fast-forwarded -- has only a single diff group).

**Comment by gerhardol** [comment]
But you also will "drop" the selected in this case so you will retain the first in the non matching commit as well. And if you have selected a file in other than the first group, it will not be selected when switching back.
I did not find a way to find when items are selected by the user and only try to retain those. (Not really a part of this issue.)

**Comment by mstv** [comment]
> But you also will "drop" the selected in this case so you will retain the first in the non matching commit as well. And if you have selected a file in other than the first group, it will not be selected when switching back.

That's true. It's another use case which should be supported, too.

> I did not find a way to find when items are selected by the user and only try to retain those. (Not really a part of this issue.)

We will need eye-tracking finally. 😉

Don't get me wrong, Gerhard. I really appreciate these features. That's why I try to get the best UX out of them.

**Comment by gerhardol** [comment]
So your suggestion is to only save and retain a file that is in the first group. Otherwise will the first file in the first group be selected for a new commit (and that file will be retained when switching again).

> Don't get me wrong, Gerhard. I really appreciate these features. That's why I try to get the best UX out of them.

This is a good discussion. I prefer a direct discussion with good arguments, that is what pushes something forward.

**Comment by gerhardol** [comment]
@mstv preview before I create a PR?

**Comment by mstv** [comment]
> preview before I create a PR?

It works as expected for my initial use case. (I have not checked with identical file names in different paths but it should work according to the code I have seen.)

Can you live with "And if you have selected a file in other than the first group, it will not be selected when switching back."? Do you have an example where it would be very useful if the file would be reselected in the other group?

BTW: How do you create and update branches like `tmp/go8`? Merging / cherry picking?

**Comment by gerhardol** [comment]
I will create a PR then

>Can you live with "And if you have selected a file in other than the first group, it will not be selected when switching back."? 

I can live with that
You had stronger opinion than me about it

> Do you have an example where it would be very useful if the file would be reselected in the other group?

I prefer to keep complexity down and to keep the retain a little better

> BTW: How do you create and update branches like tmp/go8? Merging / cherry picking?

Yes. For external like your changes, I rebase interactively and drop the current patch before applying new patches (often squashed first)
I do most dev (and private usage) with these combined feature branches, then adapt them for master.

With so many changes going on I spend too much time rebasing... 
I will do something else for some time if @RussKie still wants to review PRs

**Comment by RussKie** [comment]
> I will do something else for some time if @RussKie Igor Velikorossov FTE still wants to review PRs

Sorry, which one would you like me to review?

**Comment by gerhardol** [comment]
The big blocker is #7899, that is blocked by #7825 and #7908
There are a few other that affects the same files #7900 and #7912. #7913 has lower priority 

But #7825 and #7899 are quite big. Take your time for health, but I am blocked (or rather spending far too much time in rebasing)

**Comment by RussKie** [comment]
I've reviewed  #7908 - [one outstanding comment](https://github.com/gitextensions/gitextensions/pull/7908#discussion_r399777547), other than that it looks good.
